### ISSUE TITLE
Question on cliConfig, using a docker image in runSh

### ISSUE THREAD

**Issue by jplimpens** [Shippable/support#4148]
I have a cliConfig resource with an aws-keys integration.

I connected the cliConfig resource to a runSh job in my assembly line configuration. In my runSh job, I use the resource with scope ecr and I'm trying to pull a docker image from AWS ECR. If I understood the docs correctly an AWS CLI login should be automatically executed before custom script steps are executed

However, my job waits on:  _Please login prior to pull:_

I even added a aws-keys integration to the job, but still the same message.

The same docker image is used in CI (configured in pre_ci_boot) so I assume the integration and configured AWS policies are correct. Could you explain the difference between the CI config and the assembly line config and push me in the right direction?

**Comment by rageshkrishna** [comment]
@jplimpens Can you link us to the job that's failing so we can take a look?

**Comment by jplimpens** [comment]
@rageshkrishna How do I do that?

**Comment by rageshkrishna** [comment]
You can share the URL you see in your browser when you are looking at the job, or the ID you see in the bottom right corner of the page.

**Comment by jplimpens** [comment]
This one, I gues..  Build Id: 5a9099b53ec8c4060058f6ea

**Comment by rageshkrishna** [comment]
Thanks. We'll take a look and get back to you soon.

**Comment by rageshkrishna** [comment]
It looks like this scenario doesn't quite work out-of-the-box with `runSh` jobs right now, because the CLI authentication is actually done _inside_ of the tasks. Because your task requires authentication to pull a custom image, we are stuck in a chicken/egg situation.

As a workaround, you could try adding another task to this job that manually pulls the image. Add this task right after all of your `IN`s, and just before your existing `TASK`:

````yaml
  - TASK:
      name: pullCustomImage
      script:
        - docker pull {YOUR_CUSTOM_IMAGE_HERE}
````

This task will run inside a default container, hence allowing the CLI authentication to happen which should let the docker pull work.

I hope this helps.

**Comment by jplimpens** [comment]
Well, this workaround does fix the docker pull. However, the second task now gives me /var/lib/shippable/build/scripts/resources/cliConfig/amazonKeys/init.sh: line 30: shipctl: command not found.

Since I use a custom image the only thing I can think of is that this custom image requires some shippable specific stuff. So what do I need and how do I get it?

Build Id: 5a915ee93ec8c40600592f0c

**Comment by ambarish2012** [comment]
@jplimpens support for shipctl in custom images will be launched in our next release in approximately two weeks.

**Comment by jplimpens** [comment]
Ok, thanks. For now I'll use a workaround.

**Comment by manishas** [comment]
This has been shipped with the latest release... notes are here: https://github.com/Shippable/admiral/blob/master/releaseNotes/v6.2.4.md

**Comment by jplimpens** [comment]
Did take a while before I could play again with custom images in my runSh job. Unfortunately I still get the error `shipctl: command not found` on the step cli configuration.  Just before that on the step installing shipctl commands it gives `E: Version '1.3-1.1ubuntu1' for 'jq' was not found`.

So it seems some things are missing in my custom image. I didn't manage to find any more documentation than http://docs.shippable.com/ci/custom-docker-image/. Could you give me a list what packages must be available?

**Comment by jplimpens** [comment]
So I updated the container and to make sure I wouldn't get to much missing items I used drydock/U16 as base. I build that one from source (note it failed me on nodejs and azure cli keys) and then created my custom image. Unfortunately I'm now back at the original error `Please login prior to pull:`. AWS integration and cli are added as IN resource and workaround used as described above. What am I missing here?

**Comment by trriplejay** [comment]
hmm, it looks like you're doing everything right. I noticed that your subscription is on an older version of dynamic nodes, which uses an older version of docker (1.9).  It's possible that there's some conflict happening there between awscli and docker.

You can update to a more recent version by going to your subscription page, then click the gears in the top right, go to "node pools", then "edit" your node pool.  Select v6.3.4 from the "runtime version" dropdown. That is our most recent release.

Can you try updating that, and running your job again?

**Comment by jplimpens** [comment]
Switched to v6.3.4 and gave it a try. Same issue, different message `no basic auth credentials`. So same user, same permissions. The repo url is different. It now starts with https, contains /v2/ and manifests/1.0: But that shouldn't really matter..

So I was thinking.. after `aws ecr get-login --no-include-email` which should be executed based on the AWS integration there should run a docker login command (something like 'docker login -u AWS -p password https://aws_account_id.dkr.ecr.us-east-1.amazonaws.com` - AWS docs). But the logs give no clue on that. The logs show `Initializing scope ecr Login Succeeded` and I'm guessing this is the AWS part. Next the logs show `docker pull`. Which fails. Could it be that docker login is not executed?

**Comment by trriplejay** [comment]
@jplimpens can you verify that these same aws credentials can pull the image locally? is it possible that they don't have the correct permissions to utilize ecr?

I notice you have both a 'cliConfig' and an 'integration' resource as INs.  Do you need both? are they using the same keys? cliConfig is the one whose credentials matter, but I just wanted to make sure you weren't updating the integration resource with the correct keys while leaving the cliConfig with the incorrect ones.

**Comment by jplimpens** [comment]
Integration and cli use the same AWS Keys integration. I added both as a trial since I got stuck. 
Push and pull on the repo work fine using these credentials (just checked it again - for the I don't know how many times this day ;-) ).

Just removed the integration resource and let a new build run. See what that results in...

**Comment by jplimpens** [comment]
I created new keys and pasted these in the integration. Also I removed the workaround just to retrace my steps. Now it works.

The only thing I can think of is that the new version (6.3.4) fixed the original issue (`please login before pull`) and that the secret key has not been copied completely in the password field (which resulted in `no basic auth`).
This can be closed as far as i'm concerned.
Thanks for the help!

**Comment by trriplejay** [comment]
@jplimpens great, glad you got it working. we were having trouble coming up with any other ideas :)

**Comment by jplimpens** [comment]
Bad news.. back to `no basic auth credentials.` Worked until 8:32AM CET, after that I'm again stuck with this error. Preferably I don't want to update keys every 3 days ;-) , besides a pull with these keys work when used in a terminal. To rule out ohter things I copied the access and secret key again in the integration. But it still fails. It's not magic, so somewhere things changed, which I can't find.

Are you able to see more at your end?

**Comment by trriplejay** [comment]
Hey, it looks like you've updated your 'TASK' to run on your custom private image.  Unfortunately the step that logs in with your credentials takes place after the task image has been booted.  You can fix this by adding an additional task to your runSh job that pulls the image first before trying to use it for the task.

so your steps might look like this:
```yml
steps:
  - IN: awsCliConfig
  - TASK:
      name: pullPrivateImage
      script: #executes inside default task image, which is public
        - docker pull <myimage>:<mytag> 
  - TASK:
      name: setupStagingEnvironment
      runtime:
        options:
          imageName: <myimage>
          imageTag: <mytag>
      script: # this script will execute inside <imageName>:<imageTag>
        - echo "hello world" 

```

**Comment by jplimpens** [comment]
Well, this issue is all about using a custom private image :smiley:
Anyway, as said before I removed that workaround too (the extra task Ragesh proposed) just to retrace my steps. But, magic, magic, it just worked using the new node version and the key update. For 3 days :open_mouth: and this morning it failed again :disappointed: 

So... it's easy to add an additional task, but I still would like to understand why it worked for 3 days.
Any clues?

**Comment by trriplejay** [comment]
we've come full circle!

so, I would say that the additional task to pull the private image first is not a temporary workaround, that is a required step to run a runSh task on a private image.  The image must be pulled first inside a task that will utilize the cliConfig.

Since you're using dynamic nodes, the node will get re-used in subsequent builds until we bring the node down after a certain amount of time.  This means that after a single successful pull, subsequent jobs could docker run the private image without having to pull it again.  However, the node should not stay up for 3 days unless you were running builds non-stop, so it doesn't explain that.

As far as why you're seeing this basic auth problem again.....

I can tell you that our cliConfig is basically performing these steps (the full script is available [here](https://github.com/Shippable/execTemplates/blob/master/Ubuntu_16.04/resources/cliConfig/amazonKeys/init.sh)): 
- aws configure set (key/secret/region)
- result=$(aws ecr get-login)
- eval $result


actually.. i just noticed as i was typing this... in your cliConfig you list "eu-central-1" as the region, however the image you're trying to pull seems to be in `eu-west-1`.... if your cli is not configured in the proper region, you could see an error like this. Perhaps your local machine is configured with the correct region?

If you need to use more than one region, you can change that by adding `aws configure set region <region>` to your script, so you could do that before you pull.

Could this be the problem?

**Comment by jplimpens** [comment]
Honestly, I don't know. I'm pretty much lost. I've switched back and forth between several regions locally, and when there is no image available it pulls the image. If the image is available it returns 'image up to date'. Pretty much as expected, except we would expect an error when regions don't match... Then again... we could see that error... no guarantee we will see the error. So far so good.

Then I switched the region in the resource (the yml). Now the job doesn't fail. However, the first time after that change docker pull echoed 'image up to date' where I would have expected it would pull the image the first time (since I assume when the node gets refreshed the "local" image is lost). If the image would have been available I assume (since it worked for 3 days) I would not have ran into the error, so the 'image up to date' message  confuses me. 

In the cli the region doesn't really matter for me, so I'll leave it to west for now. However, since i don't expect the node has been 'refreshed' by now (not that I'm able to tell), and docker pull so far always says 'image up to date' the only thing I can think of is waiting for a while and see what will happen.

More to follow...
